### PR TITLE
Debugger to future transition

### PR DESCRIPTION
**Pull request by wpopielarski**
This PR is created from sschaef/debugger-to-futures branch merged with current master. May replace (but not have to) PR #993.

### CODE REVIEW

**Comment by wpopielarski** [comment]
retest this please

**Comment by sschaef** [comment]
Does this fix [#1001464](https://scala-ide-portfolio.assembla.com/spaces/scala-ide/tickets/1001464)?

**Comment by wpopielarski** [comment]
yes, it should fix it too. Also the timeouts will be diminished. There are
big now because the reason of accidental failures of tests was unknown. I
work on it tomorrow

2015-12-13 0:00 GMT+01:00 Simon Schäfer <notifications@github.com>:

**Comment by wpopielarski** [comment]
retest this please

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/async/RetainedStackManager.scala [hunk 1]
```diff
@@ -29,22 +39,30 @@ class RetainedStackManager(debugTarget: ScalaDebugTarget) extends HasLogger {
   import org.scalaide.debug.internal.launching.ScalaDebuggerConfiguration._
 
   final val MaxEntries = 20000
-  private val stackFrames: mutable.Map[ObjectReference, AsyncStackTraces] = new LRUMap(MaxEntries)
+  private val stackFrames: mutable.Map[ObjectReference, AsyncStackTraces] = {
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/async/RetainedStackManager.scala]: TODO: LRU Concurrent map? Try to find implementation

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/BreakpointSupport.scala [hunk 2]
```diff
@@ -31,30 +35,27 @@ import com.sun.jdi.event.Event
...
   final val ATTR_TYPE_NAME = "org.eclipse.jdt.debug.core.typeName"
 
-  /** Create the breakpoint support actor.
+  /**
+   * Create the breakpoint support actor.
    *
    *  @note `BreakpointSupportActor` instances are created only by the `ScalaDebugBreakpointManagerActor`, hence
    *        any uncaught exception that may occur during initialization (i.e., in `BreakpointSupportActor.apply`)
    *        will be caught by the `ScalaDebugBreakpointManagerActor` default exceptions' handler.
    */
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/BreakpointSupport.scala]: TODO: Update description

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala [hunk 3]
```diff
@@ -33,140 +39,118 @@ object ScalaDebugBreakpointManager {
...
+class ScalaDebugBreakpointManager private ( /*public field only for testing purposes */ val subordinate: ScalaDebugBreakpointSubordinate) extends IBreakpointListener {
+  /**
+   * Used to wait until all required `Future` messages have been processed.
+   *
+   * @note Use this for test purposes only!
+   */
+  private val waitForAllCurrentFutures: AtomicReference[Future[Unit]] = new AtomicReference(Future.successful {})
 
-  // from org.eclipse.debug.core.IBreakpointsListener
+  private def ravelFutures[T](b: => Future[T])(implicit ec: ExecutionContext): Unit = {
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala]: TODO: change 'ravel' to 'entangle'

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala [hunk 4]
```diff
@@ -33,140 +39,118 @@ object ScalaDebugBreakpointManager {
...
+  }
+
+  private[debug] def breakpointRequestState(breakpoint: IBreakpoint): Option[Boolean] = {
+    breakpoints.get(breakpoint).flatMap { breakpointSupport =>
+      Some(breakpointSupport.breakpointRequestState())
+    }
+  }
+
+  def reenableBreakpointsAfterHcr(changedClassesNames: Seq[String]): Future[Unit] = Future {
     /*
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala]: TODO: check if return `Future`

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/command/ScalaStepOver.scala [hunk 5]
```diff
@@ -79,90 +66,74 @@ object ScalaStepOver {
  * Actor used to manage a Scala step over. It keeps track of the request needed to perform this step.
  * This class is thread safe. Instances are not to be created outside of the ScalaStepOver object.
  */
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/command/ScalaStepOver.scala]: TODO: Update comment

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/command/ScalaStepReturn.scala [hunk 6]
```diff
@@ -27,59 +27,44 @@ object ScalaStepReturn {
  * Actor used to manage a Scala step return. It keeps track of the request needed to perform this step.
  * This class is thread safe. Instances are not to be created outside of the ScalaStepReturn object.
  */
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/command/ScalaStepReturn.scala]: TODO: Update comment

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala [hunk 7]
```diff
@@ -189,9 +173,12 @@ private class ScalaDebugBreakpointManagerActor private(debugTarget: ScalaDebugTa
       case bp: JavaLineBreakpoint if isChanged(bp.getTypeName) => bp
     }
     affectedBreakpoints.foreach { breakpoint =>
-      breakpoints(breakpoint) ! ReenableBreakpointAfterHcr
+      breakpoints(breakpoint).reenableBreakpointRequestsAfterHcr()
     }
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/breakpoints/ScalaDebugBreakpointManager.scala]: TODO: Check if this is composition of `Futures`

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugCache.scala [hunk 8]
```diff
@@ -81,7 +86,7 @@ abstract class ScalaDebugCache(val debugTarget: ScalaDebugTarget) extends HasLog
 
   /**
    * Adds the given actor as a listener for class prepare events in the debugged VM,
-   *  for types which are nested under the same outer type as the type withe the given name.
+   *  for types which are nested under the same outer type as the type with the given name.
    *  The event is sent as a ClassPrepareEvent to the actor.
    *
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugCache.scala]: TODO: Update comment

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugTarget.scala [hunk 9]
```diff
@@ -472,61 +480,82 @@ private[model] object ScalaDebugTargetActor {
...
+      }
+      findThisThread(thread)
+        .orElse(addNewThread(thread))
+        .foreach(_.suspendedFromScala(eventDetail))
+    }
+    forwardEventToRightThread(thread, eventDetail)
+  }
+
+  private[model] def updateStackFrameAfterHcr(dropAffectedFrames: Boolean): Future[Unit] = {
+    println("scala debug target update stack frame")
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugTarget.scala]: TODO: logger.info

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugTarget.scala [hunk 10]
```diff
@@ -472,61 +480,82 @@ private[model] object ScalaDebugTargetActor {
...
-      nonSystemThreads.foreach(_.updateStackFramesAfterHcr(msg))
-    case ScalaDebugTarget.ReplaceClasses(changedClasses) =>
-      replaceClassesIfVMAllows(changedClasses)
+      exit()
+      false
   }
 
-  /** Initialize this debug target actor:
+  /**
+   * Initialize this debug target actor:
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugTarget.scala]: TODO: Update comment

File: org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaJdiEventDispatcher.scala [hunk 11]
```diff
@@ -119,54 +134,34 @@ private class ScalaJdiEventDispatcherActor private (scalaDebugTargetActor: Suppr
    * Resume or not the stopped threads depending on actor's answers
    */
   private def processEventSet(eventSet: EventSet): Unit = {
+    import scala.concurrent.ExecutionContext.Implicits._
     import scala.collection.JavaConverters._
 
-    var futures = List[Future[Any]]()
-
+    var staySuspendeds = List[Future[Boolean]]()
+eventSet.asScala.foreach { println }
```
Comment (wpopielarski) [inline org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaJdiEventDispatcher.scala]: TODO: `logger.debug`

**Comment by wpopielarski** [comment]
retest this please

File: org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala [hunk 12]
```diff
@@ -208,9 +219,9 @@ class RemoteConnectorTest {
   /**
    * Check if it is possible to connect to a running VM that did not suspend.
    */
-  @Test(timeout = 5000)
+  @Ignore("Debugee cannot guarantee to wait for its Debugging")
+  @Test(timeout = 10000)
```
Comment (sschaef) [inline org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala]: The configured timeout is still huge. Do you want to keep them this way?

Comment (wpopielarski) [inline org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala]: This is for this ignored test, but I'll change it. Locally it does not need
it so long

2015-12-29 12:44 GMT+01:00 Simon Schäfer <notifications@github.com>:

Comment (sschaef) [inline org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/RemoteConnectorTest.scala]: Indeed, I missed the ignore.

**Comment by sschaef** [comment]
LGTM

**Pull request merged**